q2_quality_filter/_filter.py

```diff
@@ -6,181 +6,448 @@
+        The PHRED offset encoding of the quality scores.
+    min_quality : int
+        The minimum quality that a base must have in order to not be considered
+        part of a low quality window.
+    window_length : int
+        The length of the low quality window to search for.
+
+    Returns
+    -------
+    int or None
+        The index of the first position of the first low quality window found
+        or None if no such window is found.
+
+    '''
+    # parse and adjust quality scores
+    quality_scores_parsed = np.frombuffer(
+        quality_scores, np.uint8
+    )
+    quality_scores_adjusted = quality_scores_parsed - phred_offset
+    less_than_min_quality = quality_scores_adjusted < min_quality
+
+    # use a convolution to detect bad quality windows
+    window = np.ones(window_length, dtype=int)
+    convolution = np.convolve(less_than_min_quality, window, mode='valid')
+    window_indices = np.where(convolution == window_length)[0]
+
+    if len(window_indices) == 0:
+        return None
+
+    return window_indices[0]
+
+
+def _truncate(fastq_record: FastqRecord, position: int) -> FastqRecord:
+    '''
+    Truncates a fastq record's sequence and quality scores to a specified
+    `position`. Note that `position` is the first position that is excluded
+    from the resulting record.
+
+    Parameters
+    ----------
+    fastq_record : FastqRecord
+        The fastq record to truncate
+    position : int
+        The truncation position
+
+    Returns
+    -------
+    FastqRecord
+        The truncated fastq record.
+    '''
+    fastq_record.sequence = fastq_record.sequence[:position]
+    fastq_record.quality_scores = fastq_record.quality_scores[:position]
+
+    return fastq_record
+
+
+class RecordStatus(Enum):
+    UNTRUNCATED = 1
+    TRUNCATED = 2
+    SHORT = 3
+    AMBIGUOUS = 4
+    TRUNCATED_AMBIGUOUS = 5
+
+
+def _process_record(
+    fastq_record: FastqRecord,
+    phred_offset: int,
+    min_quality: int,
+    window_length: int,
+    min_length_fraction: float,
+    max_ambiguous: int,
+) -> tuple[FastqRecord, RecordStatus]:
+    '''
+    Processes a fastq record by detecting low quality windows, truncating if
+    one or more such windows are found, detecting if a truncated record is too
+    short, and finally detecting if the number of ambiguous bases is too high.
+
+    Parameters
+    ----------
+    fastq_record : FastqRecord | None
+        The fastq record to be processed. None if record does not exist (for
+        convenience when reverse reads are not present).
+    phred_offset : int
+        The PHRED encoding of the record's quality scores.
+    min_quality : int
+        The minimum quality that a base must have in order to not be considered
+        part of a low quality window.
+    window_length : int
+        The length of the low quality window to search for.
+    min_length_fraction : float
+        The fraction of its original length a record must be greater than to
```

Comment:
This `min` is interpreted differently than `min_quality`. Is that how it was before? High quality would be anything  >= min_quality, but long enough would be anything > min_length_fraction. It's documented correctly here, so if that's how it was before it's probably safest just to leave it, but if this is newly introduced we should make these consistent (I would say long enough >= min_length_fraction makes the most sense).

Reply:
yes, this is how it was before